drivers/i2c/i2c_gpio.c

```diff
@@ -141,16 +148,32 @@ static int i2c_gpio_init(struct device *dev)
 		return -EINVAL;
 	}
 
-	err = gpio_config(context->sda_gpio, config->sda_pin,
-			  config->sda_flags | GPIO_INPUT | GPIO_OUTPUT_HIGH);
-	if (err == -ENOTSUP) {
+	gpio_flags_t flags = config->sda_flags;
+
+	if (flags & GPIO_SINGLE_ENDED){
```

Comment:
What gpio peripheral does not support open drain?  Does it properly return `-ENOTSUP` when asked to configure open drain?

Can the problem be avoided by not configuring the signals as `GPIO_OPEN_DRAIN` in devicetree?

Reply:
The initial idea of this commit was exactly to work if open drain not configured in device tree. Later I added a capability to fall back to this mode if open drain is not supported.

Reply:
I believe it's reasonable to require that the GPIO flags be correctly configured in devicetree for the hardware that uses the driver and devicetree.

Are any of these changes necessary under that expectation?

Reply:
If we expect that the device tree configuration is correct we can remove the error check and just fall back to explicit direction switching if GPIO_SINGLE_ENDED flag not provided.

Reply:
I'm not sure I follow.  I would still expect the code to be:

```
err = gpio_config(context->sda_gpio, config->sda_pin,
￼			  config->sda_flags | GPIO_INPUT | GPIO_OUTPUT_HIGH);
if (err == -ENOTSUP) {
    // configure for direction control here: confirm both GPIO_INPUT and GPIO_OUTPUT configurations are 
    // acceptable.
}
```

I think support for systems that require direction switching is fine, but it should be based on having the GPIO driver validate the flags by being asked to configure the pin in the two modes required, rather than have the bitbang driver interpret them.

Reply:
So you propose to test if the driver allows to configure the pin both as input and as output when switching to the newly introduced mode, right?

Reply:
Yes, basically do exactly what will be done in the SDA control function.

Reply:
I updated the commit following the recommendations above.

Reply:
That isn't quite what I thought I was requesting, but I'm not sure what I have in mind is correct.

What specific platform(s) do you want this for, that cannot be made to work with the existing code and appropriate devicetree flags?  I need to do some experimentation.

Reply:
Actually I found a problem with i2c bitbang on NRF52 but it was before commit [b9e3226f](url=https://github.com/zephyrproject-rtos/zephyr/commit/b9e3226f6042e018de42ac40f3a2603243731fa7) . My idea was to switch the SDA direction explicitly if GPIO_SINGLE_ENDED not configured or not working properly. But before pull request I had to rebase, and found that now both solutions work fine: mine and yours. Still I suppose my fix will work on any platform which does not support open drain or even does not support simultaneous operation of pin as input and output, so it has sense.

Reply:
For posterity: what I was looking for was something more like this, which pushes flag validation down to the GPIO controller.  More explanation in the review.

```
diff --git a/drivers/i2c/i2c_gpio.c b/drivers/i2c/i2c_gpio.c
index dccbdc9c6a..edbbeee16c 100644
--- a/drivers/i2c/i2c_gpio.c
+++ b/drivers/i2c/i2c_gpio.c
@@ -55,6 +55,7 @@ struct i2c_gpio_context {
        const struct device *sda_gpio;  /* GPIO used for I2C SDA line */
        gpio_pin_t scl_pin;             /* Pin on gpio used for SCL line */
        gpio_pin_t sda_pin;             /* Pin on gpio used for SDA line */
+       bool push_pull;                 /* SDA must be configured for direction */
 };
 
 static void i2c_gpio_set_scl(void *io_context, int state)
@@ -143,10 +144,20 @@ static int i2c_gpio_init(const struct device *dev)
 
        err = gpio_config(context->sda_gpio, config->sda_pin,
                          config->sda_flags | GPIO_INPUT | GPIO_OUTPUT_HIGH);
-       if (err == -ENOTSUP) {
+       if ((err == -ENOTSUP)
+           && (GPIO_SINGLE_ENDED & config->sda_flags)) {
                err = gpio_config(context->sda_gpio, config->sda_pin,
                                  config->sda_flags | GPIO_OUTPUT_HIGH);
        }
+       if (err == -ENOTSUP) {
+               config->push_pull = true;
+               err = gpio_config(context->sda_gpio, config->sda_pin,
+                                 config->sda_flags | GPIO_INPUT);  /* confirms flags are compatible with input */
+               if (err == 0) {
+                       err = gpio_config(context->sda_gpio, config->sda_pin,
+                                         config->sda_flags | GPIO_OUTPUT_HIGH);
+               }
+       }
        if (err) {
                LOG_ERR("failed to configure SDA GPIO pin (err %d)", err);
                return err;
```

Reply:
Got your idea. I propose a more compromissive variant. Fall back to explicit switching if GPIO_SINGLE_ENDED not configured or if SINGLE_ENDED mode can not be configured. Verify both OUTPUT_HIGH and INPUT modes.

Reply:
No.  I don't want the first test to be the i2c_gpio driver assuming it knows whether the GPIO controller needs the `GPIO_SINGLE_ENDED` flag set in order to configure bidirectional signals.  If the proposal I gave above doesn't work I need to understand why.  Only if it gets into the branch where `push_pull` is set do we need the SDA `gpio_dt_flags` added to the context, and we should use the `push_pull` flag rather than gpio pin flags to determine what `i2c_gpio_{s,g}et_sda` do.

Also this still would need massive amounts of testing.  So I remain unconvinced it's worth doing this now.

Reply:
The proposal above is workable, still the original idea is to fall back to push-pull mode if GPIO_SINGLE_ENDED not provided in device tree or if applying the GPIO_SINGLE_ENDED configuration failed. So you can just specify flags "0" in device tree ad force push-pull mode with explicit direction switching. Or the push-pull mode can be enabled if SINGLE_ENDED configuration can not be applied.